__NOTE:__

1) This document is arranged in an order following the __Case_description__

2) Section A OVERVIEW is available in the __Case_description__

3) This document starts from SECTION B  DATA PREPARATION as per __Case_desription__ 


In [1]:
#Import libraries 
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt 
import seaborn as sns
import datetime
%matplotlib inline 

# SECTION B     DATA PREPARATION
##  B.2         Data pre-processing

### B.2.1   Item Data 

In [2]:
#read in items data
items = pd.read_csv('Item.csv', sep = ',')

In [3]:
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 6 columns):
ITEM             int64
CATEGORY         object
COLOR            object
SUPLID           object
PURCHASEPRICE    float64
SALEPRICE        int64
dtypes: float64(1), int64(2), object(3)
memory usage: 100.2+ MB


In [150]:
items.PURCHASEPRICE.isnull().any()

False

#### Data-type contraints: cleaning, verifying and reporting

In [4]:
# Change dtype of saleprice to decimal 
before = type(items.SALEPRICE[0])
items.SALEPRICE = items.SALEPRICE.astype('float64')
after = type(items.SALEPRICE[0])
before
after

numpy.float64

In [5]:
# check the size of the dataframe 
items.shape

(2188800, 6)

In [6]:
# check the first 5 rows of the items dataframe 
items.head()

ITEM CATEGORY   COLOR   SUPLID  PURCHASEPRICE  SALEPRICE
0  127521  PAJAMAS    PINK  41560EE        1867.09     2800.0
1  151662    SOCKS    GOLD  16091ZZ        1862.08     2050.0
2  453756     COAT    GOLD  11370LL         914.81     1370.0
3  434386  PAJAMAS  SILVER  41560EE        1913.00     1430.0
4  422712    DRESS  SILVER  48094JJ         851.43      940.0

In [7]:
# check data types of the columns based on the data dictionary 
items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 6 columns):
ITEM             int64
CATEGORY         object
COLOR            object
SUPLID           object
PURCHASEPRICE    float64
SALEPRICE        float64
dtypes: float64(2), int64(1), object(3)
memory usage: 100.2+ MB


##### Reporting: As seen above, the currently stored items data type meets the data-type contraints requirments. 

### B.2.2    Customer Data 

In [10]:
#read in customer data
customers = pd.read_csv('CustomerData.txt', sep='|')
customers

USERID GENDER         DOB               COUNTRY              EDUCATION  \
0     39060      F  1982-09-18              Ethiopia                          
1     26050      M  1988-11-22  United Arab Emirates                          
2     48580      M  1975-12-06            Martinique               Bachelor   
3     40220      F  1984-11-16               Germany                 Master   
4     25160      M  1977-12-05            Madagascar                 Master   
5     37700      M  1972-12-19                Guinea            High school   
6     25080      M  1987-03-03                 Japan               Bachelor   
7     34140      N  1972-06-11               Algeria            High school   
8     27260      F  2000-12-28           Puerto Rico            High school   
9     37480      M  1966-04-09            Antarctica                          
10    19140      F  1994-06-14               Senegal                 Master   
11    17410      F  1978-03-30                France                 Master   
12    20880      F  1978-03-20                 Chile                 Master   
13    49710      F  1968-09-12                 Japan                 Master   
14    27760      F  1980-02-23             Gibraltar                          
15    48100      F  1973-04-25               Reunion                 Master   
16    40240      F  1990-07-01            Martinique                          
17    29220      F  1993-09-26               Morocco            High school   
18    38130      F  1985-06-09                Mexico               Bachelor   
19    12670      F  1982-05-03               Uruguay            High school   
20    48550      M  1975-02-19                France               Bachelor   
21    48520      M  1983-02-20             Gibraltar            High school   
22    43290      F  1984-08-16              Guernsey            High school   
23    27120      F  1989-12-15             Macedonia            High school   
24    17820      F  1980-07-01               Liberia                          
25    31310      N  1971-09-06                Jordan                          
26    13240      F  1993-12-30                 Malta               Bachelor   
27    42330      F  1993-07-27               Bahamas               Bachelor   
28    46530      F  1967-03-24                 Japan                          
29    40380      M  1971-01-11             Greenland               Bachelor   
..      ...    ...         ...                   ...                    ...   
544   45260      F  1993-08-12               Uruguay               Bachelor   
545   26310      F  1974-11-30              Guernsey               Bachelor   
546   35330      N  1990-12-18               Andorra                 Master   
547   49000      N  1965-09-26               Senegal  Less than high school   
548   48470      M  1986-08-11               Romania  Less than high school   
549   38480      M  1989-10-15               Croatia  Less than high school   
550   28280      F  1983-01-24               Reunion                 Master   
551   39420      M  1995-12-25               Senegal  Less than high school   
552   26410      M  1967-09-01        United Kingdom  Less than high school   
553   33720      F  1983-11-23               Liberia               Bachelor   
554   23890      M  1965-08-11               Belgium            High school   
555   31450      M  1974-07-28             Macedonia                          
556   24690      F  1969-03-21               Moldova            High school   
557   30050      M  1969-10-18               Belgium                 Master   
558   40940      F  1981-05-17             Greenland            High school   
559   44460      M  1973-11-17              Thailand            High school   
560   30720      F  1967-05-11                Cyprus                 Master   
561   36140      M  1967-06-04             Gibraltar               Bachelor   
562   23910      M  1967-05-10            Mozambique   

In [11]:
#check dtypes 
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 574 entries, 0 to 573
Data columns (total 6 columns):
USERID       574 non-null int64
GENDER       574 non-null object
DOB          574 non-null object
COUNTRY      574 non-null object
EDUCATION    574 non-null object
HOBBY        574 non-null object
dtypes: int64(1), object(5)
memory usage: 27.0+ KB


#### Data-type contraints: cleaning, verifying and reporting

In [12]:
#change "DOB" field to date dtype

customers['DOB']= pd.to_datetime(customers['DOB']) 

In [13]:
#check dtype again
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 574 entries, 0 to 573
Data columns (total 6 columns):
USERID       574 non-null int64
GENDER       574 non-null object
DOB          574 non-null datetime64[ns]
COUNTRY      574 non-null object
EDUCATION    574 non-null object
HOBBY        574 non-null object
dtypes: datetime64[ns](1), int64(1), object(4)
memory usage: 27.0+ KB


##### Reporting: As seen above, the currently stored customers data type meets the data-type contraints requirments.

In [14]:
# save a copy of the currently stored data in .csv file. 
customers_prep = pd.DataFrame(customers)
customers_prep.to_csv('customers_prep.csv')
customers_prep.to_csv(r'C:\Users\YUJI6004\customers_prep.csv', index=False)

### B.2.3  Transaction Data 

In [15]:
#read in transaction data 
transactions = pd.read_csv('TransactionsData.txt', sep='|', low_memory=False)
transactions

USERID WEBBROWSER   PPC_ADD    ITEM PURCHASE QTY DISCOUNT  \
0        15410     CHROME     KEVIN  309261      YES   4       28   
1        33510   EXPLORER     SENDI  484592      YES   4       30   
2        26390  MOBILEAPP     KEVIN  191104      YES   2        4   
3        48580  MOBILEAPP     SENDI  124180      YES   3       29   
4        14600  MOBILEAPP     KEVIN  223591      YES            7   
5        43930  MOBILEAPP  ISABELLA  147957      YES   2        4   
6        18630     SAFARI  ISABELLA  467237      YES   2       11   
7        44480   EXPLORER     SENDI  213633       NO   2       30   
8        41980     CHROME        NO  211943       NO   2       17   
9        13270     CHROME        NO  417382      YES   2       21   
10       12170     SAFARI     KEVIN  496435      YES   2        7   
11       18970  MOBILEAPP            473750      YES   2       14   
12       43800     SAFARI        NO  224072      YES   2       30   
13       42570   EXPLORER     KEVIN  233718      YES   2       14   
14       10980  MOBILEAPP  ISABELLA  459437      YES   4       23   
15       47230  MOBILEAPP     KEVIN  247537      YES   2       17   
16       16820     CHROME    BIANCA  370283      YES   4        7   
17       39960     CHROME     KEVIN  355944            2       17   
18       29590     CHROME    BIANCA  278217      YES   3       28   
19       21430     CHROME  ISABELLA  374287      YES   2       10   
20       18630  MOBILEAPP        NO  413157      YES   4        8   
21       38080     CHROME     KEVIN  430226      YES   2       11   
22       14720     SAFARI        NO  290775      YES   2       24   
23       12900   EXPLORER        NO  320662            2       15   
24       36950   EXPLORER     KEVIN  433242      YES   3        4   
25       32580               BIANCA  486100      YES   2       24   
26       36770  MOBILEAPP        NO  358635            3       11   
27       35050     CHROME     SENDI  108749      YES   2        8   
28       33470   EXPLORER    BIANCA  176687      YES   2       29   
29       19580     CHROME     KEVIN  296196      YES   3        2   
...        ...        ...       ...     ...      ...  ..      ...   
2188770  46390  MOBILEAPP            349730      YES   3       24   
2188771  35460   EXPLORER    BIANCA  282832       NO   2        8   
2188772  25680     SAFARI  ISABELLA  247797       NO   3        6   
2188773  11410             ISABELLA  430707       NO   3       22   
2188774  43460   EXPLORER     KEVIN  369607      YES   4       28   
2188775  21350     CHROME  ISABELLA  448712       NO   3       20   
2188776  49310     SAFARI    BIANCA  328098      YES   2       19   
2188777  27620  MOBILEAPP        NO  213880            4       23   
2188778  16170  MOBILEAPP    BIANCA  493055      YES   2        8   
2188779  16720     CHROME     KEVIN  183642      YES   2        7   
2188780  35330     SAFARI    BIANCA  341696      YES   3       21   
2188781  19270     SAFARI        NO  456252      YES   4       18   
2188782  25360   EXPLORER     KEVIN  451312      YES   2       29   
2188783  39930  MOBILEAPP     SENDI  254063            3       23   
2188784  36740  MOBILEAPP  ISABELLA  247056      YES   2       16   
2188785  24340     CHROME    BIANCA  310769      YES   3        8   
2188786  36770     CHROME        NO  209967      YES   2       26   
2188787  43950     CHROME    BIANCA  378460      YES   3        6   
2188788  33940  MOBILEAPP     SENDI  313850      YES            6   
2188789  14270     CHROME        NO  404967      YES   2       10   
2188790  24340  MOBILEAPP     SENDI  461738       NO   2        1   
2188791  37500     CHROME     KEVIN  351901      YES   3       13   
2188792  24250                   NO  218937      YES   2       17   
2188793  29060  MOBILEAPP  ISABELLA  460633      YES   3       13   
2188794  39790     CHROME     KENZA  138272      YES   3       16   
2188795  14500   EXPLORER     KENZA  387248      YES   3        5   
218879

#### Data-type contraints: cleaning, verifying and reporting

In [16]:
# check the raw data type and compare with data dictionary 
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          object
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             object
DISCOUNT        object
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        object
DELIVERYDATE    object
REVIEW          object
RATING          object
TRACKNO         object
TIMESTAMP       object
dtypes: int64(1), object(14)
memory usage: 250.5+ MB


#### Correct the following datatypes in transactions DB: 
Rating(int64),TIMESTAMP(Time), UserID (int64 ), QTY(int64), Discount (int64),  SHIPDAYS(int64), DELIVERYDATE (Date),  TRACKNO(int64)

#### 1) Change "RATING" from object to int64

In [17]:
#Check unique values in RATING 
transactions.RATING.unique()

array(['1', '4', ' ', '3', '2', '5'], dtype=object)

In [18]:
# replace ' ' with 0 in RATING field 
transactions['RATING'].replace(to_replace = ' ',value ="0", inplace=True) 

In [19]:
# Re-check unique values in RATING after replacement 
transactions['RATING'].unique()

array(['1', '4', '0', '3', '2', '5'], dtype=object)

In [20]:
# change data type 
transactions['RATING'] = transactions['RATING'].astype('int64')

In [21]:
# check datatype 
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          object
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             object
DISCOUNT        object
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        object
DELIVERYDATE    object
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       object
dtypes: int64(2), object(13)
memory usage: 250.5+ MB


#### 2) Change "TIMESTAMP" from object to timestamp

In [22]:
# check unique values in TIMESTAMP 
transactions['TIMESTAMP'].unique()

array(['21:54:45', '5:44:25', '6:58:15', '12:50:23', '5:23:37', '2:47:2',
       '2:31:39', '7:56:56', '0:31:21', '12:18:21', '12:48:46', '19:2:37',
       '1:26:31', '23:7:53', '15:18:22', '1:18:40', '0:3:15', '11:56:58',
       '11:9:59', '12:16:2', '1:31:40', '20:11:54', '7:2:59', '12:9:28',
       '12:7:37', '10:30:9', '0:28:54', '9:2:29', '11:23:13', '7:5:34',
       '12:3:51', '13:9:53', '23:41:5', '20:18:53', '13:24:6', '13:57:21',
       '12:25:54', '2:30:52', '1:45:36', '20:16:1', '9:54:58', '5:50:52',
       ' ', '9:42:18', '9:9:58', '0:44:26', '21:9:24', '9:42:39',
       '1:47:18', '12:25:23', '23:16:25', '5:56:53', '11:50:25',
       '19:34:14', '10:0:5', '6:16:2', '1:13:51', '5:55:3', '2:8:26',
       '15:20:4'], dtype=object)

In [23]:
#replace unknown timestamp (' ') with "00:00:00"
transactions['TIMESTAMP'].replace(to_replace = ' ',value ="00:00:00", inplace=True) 

In [24]:
# re-check unique values in TIMESTAMP
transactions['TIMESTAMP'].unique()

array(['21:54:45', '5:44:25', '6:58:15', '12:50:23', '5:23:37', '2:47:2',
       '2:31:39', '7:56:56', '0:31:21', '12:18:21', '12:48:46', '19:2:37',
       '1:26:31', '23:7:53', '15:18:22', '1:18:40', '0:3:15', '11:56:58',
       '11:9:59', '12:16:2', '1:31:40', '20:11:54', '7:2:59', '12:9:28',
       '12:7:37', '10:30:9', '0:28:54', '9:2:29', '11:23:13', '7:5:34',
       '12:3:51', '13:9:53', '23:41:5', '20:18:53', '13:24:6', '13:57:21',
       '12:25:54', '2:30:52', '1:45:36', '20:16:1', '9:54:58', '5:50:52',
       '00:00:00', '9:42:18', '9:9:58', '0:44:26', '21:9:24', '9:42:39',
       '1:47:18', '12:25:23', '23:16:25', '5:56:53', '11:50:25',
       '19:34:14', '10:0:5', '6:16:2', '1:13:51', '5:55:3', '2:8:26',
       '15:20:4'], dtype=object)

In [25]:
transactions['TIMESTAMP'] = pd.to_timedelta(transactions['TIMESTAMP'])

In [26]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          object
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             object
DISCOUNT        object
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        object
DELIVERYDATE    object
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       timedelta64[ns]
dtypes: int64(2), object(12), timedelta64[ns](1)
memory usage: 250.5+ MB


In [27]:
transactions.TIMESTAMP.unique()

array([78885000000000, 20665000000000, 25095000000000, 46223000000000,
       19417000000000, 10022000000000,  9099000000000, 28616000000000,
        1881000000000, 44301000000000, 46126000000000, 68557000000000,
        5191000000000, 83273000000000, 55102000000000,  4720000000000,
         195000000000, 43018000000000, 40199000000000, 44162000000000,
        5500000000000, 72714000000000, 25379000000000, 43768000000000,
       43657000000000, 37809000000000,  1734000000000, 32549000000000,
       40993000000000, 25534000000000, 43431000000000, 47393000000000,
       85265000000000, 73133000000000, 48246000000000, 50241000000000,
       44754000000000,  9052000000000,  6336000000000, 72961000000000,
       35698000000000, 21052000000000,              0, 34938000000000,
       32998000000000,  2666000000000, 76164000000000, 34959000000000,
        6438000000000, 44723000000000, 83785000000000, 21413000000000,
       42625000000000, 70454000000000, 36005000000000, 22562000000000,
      

#### 3) Change "USERID" from object to int64

In [28]:
transactions.USERID.unique()

array(['15410', '33510', '26390', '48580', '14600', '43930', '18630',
       '44480', '41980', '13270', '12170', '18970', '43800', '42570',
       '10980', '47230', '16820', '39960', '29590', '21430', '38080',
       '14720', '12900', '36950', '32580', '36770', '35050', '33470',
       '19580', '13940', '13560', '45260', '32950', '14480', '45590',
       '16040', '42920', '28250', '48760', '14940', '36740', '41930',
       '26970', '14440', '11080', '31950', '26930', '14070', '29310',
       '37500', '42170', '43600', ' ', '14790', '30360', '42980', '40380',
       '26130', '27320', '25630', '43290', '44620', '12130', '21830',
       '23820', '12420', '38860', '31750', '40060', '28900', '27550',
       '16610', '32340', '29070', '20760', '18490', '36980', '21960',
       '39400', '24160', '11630', '31340', '16360', '33720', '37700',
       '23520', '15770', '40810', '43920', '28650', '35110', '40340',
       '19420', '34140', '31310', '39360', '36180', '38130', '23450',
       '43540',

In [29]:
transactions.USERID.max()

'49980'

In [30]:
#replace unknown USERID (' ') with "99999"
transactions['USERID'].replace(to_replace = ' ',value ='50000', inplace = True) 

In [31]:
transactions.USERID = transactions.USERID.astype('int64')

In [32]:
transactions.USERID.dtype

dtype('int64')

#### 4) Change "QTY" from object to int64

In [33]:
transactions.QTY = transactions.USERID.astype('int64')

In [34]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        object
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        object
DELIVERYDATE    object
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       timedelta64[ns]
dtypes: int64(4), object(10), timedelta64[ns](1)
memory usage: 250.5+ MB


#### 5) Change "DISCOUNT" from object to int64

In [35]:
transactions.DISCOUNT.unique()

array(['28', '30', '4', '29', '7', '11', '17', '21', '14', '23', '10',
       '8', '24', '15', '2', '5', '20', '19', '26', '1', '25', '18', '13',
       '6', '16', '27', '22', ' ', '3', '9'], dtype=object)

In [36]:
transactions.DISCOUNT.replace(to_replace = ' ',value ='00', inplace = True) 

In [37]:
transactions.DISCOUNT.unique()

array(['28', '30', '4', '29', '7', '11', '17', '21', '14', '23', '10',
       '8', '24', '15', '2', '5', '20', '19', '26', '1', '25', '18', '13',
       '6', '16', '27', '22', '00', '3', '9'], dtype=object)

In [38]:
transactions.DISCOUNT = transactions.DISCOUNT.astype('int64')

In [39]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        int64
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        object
DELIVERYDATE    object
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       timedelta64[ns]
dtypes: int64(5), object(9), timedelta64[ns](1)
memory usage: 250.5+ MB


#### 6) Change "SHIPDAYS " from object to int64

In [40]:
transactions.SHIPDAYS.unique() 

array(['6', '5', '7', '1', ' ', '3', '10', '8', '2', '4'], dtype=object)

In [41]:
transactions.SHIPDAYS.replace(to_replace = ' ',value ='00', inplace = True) 

In [42]:
transactions.SHIPDAYS = transactions.SHIPDAYS.astype('int64')

In [43]:
transactions.SHIPDAYS.dtype

dtype('int64')

In [44]:
transactions.SHIPDAYS.unique()

array([ 6,  5,  7,  1,  0,  3, 10,  8,  2,  4], dtype=int64)

In [45]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        int64
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        int64
DELIVERYDATE    object
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       timedelta64[ns]
dtypes: int64(6), object(8), timedelta64[ns](1)
memory usage: 250.5+ MB


#### 7) Change "DELIVERYDATE " from object to date

In [46]:
transactions['DELIVERYDATE'].unique()

array(['2018-5-25', '2018-5-1', '2018-7-7', '2018-2-13', '2018-1-10',
       '2018-7-20', '2018-11-21', '2018-2-6', '2018-3-26', '2018-10-3',
       '2018-10-7', '2018-5-14', '2018-5-20', '2018-3-24', '2018-4-24',
       '2018-12-8', '2018-1-27', '2018-10-21', '2018-10-19', '2018-3-13',
       '2018-5-17', '2018-4-23', '2018-12-19', '2018-11-23', '2018-3-28',
       '2018-7-3', '2018-1-13', ' '], dtype=object)

In [47]:
transactions.DELIVERYDATE.replace(to_replace = ' ',value ='2019-12-30', inplace = True) 

In [48]:
transactions['DELIVERYDATE'].unique()

array(['2018-5-25', '2018-5-1', '2018-7-7', '2018-2-13', '2018-1-10',
       '2018-7-20', '2018-11-21', '2018-2-6', '2018-3-26', '2018-10-3',
       '2018-10-7', '2018-5-14', '2018-5-20', '2018-3-24', '2018-4-24',
       '2018-12-8', '2018-1-27', '2018-10-21', '2018-10-19', '2018-3-13',
       '2018-5-17', '2018-4-23', '2018-12-19', '2018-11-23', '2018-3-28',
       '2018-7-3', '2018-1-13', '2019-12-30'], dtype=object)

In [49]:
transactions['DELIVERYDATE']= pd.to_datetime(transactions['DELIVERYDATE']) 

In [50]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        int64
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        int64
DELIVERYDATE    datetime64[ns]
REVIEW          object
RATING          int64
TRACKNO         object
TIMESTAMP       timedelta64[ns]
dtypes: datetime64[ns](1), int64(6), object(7), timedelta64[ns](1)
memory usage: 250.5+ MB


In [51]:
transactions.head()

USERID WEBBROWSER PPC_ADD    ITEM PURCHASE    QTY  DISCOUNT    PAYMENT  \
0   15410     CHROME   KEVIN  309261      YES  15410        28  APPLE PAY   
1   33510   EXPLORER   SENDI  484592      YES  33510        30              
2   26390  MOBILEAPP   KEVIN  191104      YES  26390         4              
3   48580  MOBILEAPP   SENDI  124180      YES  48580        29       AMEX   
4   14600  MOBILEAPP   KEVIN  223591      YES  14600         7  APPLE PAY   

  WAREHOUSE  SHIPDAYS DELIVERYDATE REVIEW  RATING TRACKNO TIMESTAMP  
0         D         6   2018-05-25     NO       1  348721  21:54:45  
1         D         5   2018-05-01     NO       4  373987  05:44:25  
2         B         7   2018-07-07    YES       0  421013  06:58:15  
3         E         5   2018-02-13     NO       3  406643  12:50:23  
4         B         6   2018-01-10     NO       0  365623  05:23:37

#### 8) Change "TRACKNO" from object to int64

In [52]:
transactions.TRACKNO.unique()

array(['348721', '373987', '421013', ..., '403626', '477380', '320491'],
      dtype=object)

In [53]:
transactions.TRACKNO.max()

'4e+05'

In [54]:
transactions.TRACKNO.replace(to_replace = '3e+05',value ='300000', inplace = True) 
transactions.TRACKNO.replace(to_replace = '4e+05',value ='400000', inplace = True) 
transactions.TRACKNO.replace(to_replace = ' ',value ='500000', inplace = True) 

In [55]:
transactions.TRACKNO = transactions.TRACKNO.astype('int64')

#### Validate date types and data display 

In [56]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        int64
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        int64
DELIVERYDATE    datetime64[ns]
REVIEW          object
RATING          int64
TRACKNO         int64
TIMESTAMP       timedelta64[ns]
dtypes: datetime64[ns](1), int64(7), object(6), timedelta64[ns](1)
memory usage: 250.5+ MB


In [57]:
transactions.head()

USERID WEBBROWSER PPC_ADD    ITEM PURCHASE    QTY  DISCOUNT    PAYMENT  \
0   15410     CHROME   KEVIN  309261      YES  15410        28  APPLE PAY   
1   33510   EXPLORER   SENDI  484592      YES  33510        30              
2   26390  MOBILEAPP   KEVIN  191104      YES  26390         4              
3   48580  MOBILEAPP   SENDI  124180      YES  48580        29       AMEX   
4   14600  MOBILEAPP   KEVIN  223591      YES  14600         7  APPLE PAY   

  WAREHOUSE  SHIPDAYS DELIVERYDATE REVIEW  RATING  TRACKNO TIMESTAMP  
0         D         6   2018-05-25     NO       1   348721  21:54:45  
1         D         5   2018-05-01     NO       4   373987  05:44:25  
2         B         7   2018-07-07    YES       0   421013  06:58:15  
3         E         5   2018-02-13     NO       3   406643  12:50:23  
4         B         6   2018-01-10     NO       0   365623  05:23:37

##### Reporting: As seen above, the currently stored transactions data type meets the data-type contraints requirments.

In [58]:
# save a copy of the currently stored transaction database as .csv
transactions_prep = pd.DataFrame(transactions)
transactions_prep.to_csv('transactions_prep.csv')
transactions_prep.to_csv(r'C:\Users\YUJI6004\transactions_prep.csv', index=False)

## B.3   Data Cleaning

### B.3.1  Data Quality 
#### Items Data

#### 1) Check for duplicated rows

In [291]:
items_prep.duplicated(['ITEM', 'CATEGORY', 'COLOR', 'SUPLID', 'PURCHASEPRICE', 'SALEPRICE']).value_counts()

True     1904410
False     284390
dtype: int64

In [178]:
items_prep.duplicated(['ITEM', 'CATEGORY', 'COLOR', 'SUPLID', 'PURCHASEPRICE', 'SALEPRICE']).value_counts()*100/len(items_prep)

True     87.007036
False    12.992964
dtype: float64

#### Reporting: as seen above, almost 87% of data are duplicates across all fields. We need to remove duplicates. 

In [295]:
# drop duplicates 
items_prep.drop_duplicates(inplace=True)

In [296]:
# verify if duplicates are dropped 
items_prep.duplicated(['ITEM', 'CATEGORY', 'COLOR', 'SUPLID', 'PURCHASEPRICE', 'SALEPRICE']).value_counts()

False    284390
dtype: int64

#### 1) Check Range Constraints: ITEM, CATEGORY, COLOR

In [297]:
# Let's start with quantitative variables 
items_prep.describe()

ITEM  PURCHASEPRICE      SALEPRICE
count  284390.000000  284390.000000  284390.000000
mean   298372.001062    1001.057978    1343.938113
std    115279.872551     577.360400     842.701588
min    100013.000000       0.000000       0.000000
25%    199528.000000     494.090000     640.000000
50%    296885.000000    1007.460000    1280.000000
75%    397557.000000    1504.110000    1980.000000
max    499997.000000    1999.700000    3600.000000

In [314]:
items_prep.ITEM.nunique()

10944

##### As seen above, item is within [100000, 500000] range. 

In [298]:
# display disctinct value in COLOR
items_prep.COLOR.unique()

array(['PINK', 'GOLD', 'SILVER', 'RAINBOW', 'VIOLET'], dtype=object)

In [299]:
# display unique values in CATEGORY
items_prep.CATEGORY.unique()

array(['PAJAMAS', 'SOCKS', 'COAT', 'DRESS', 'SCARF', 'BELT', 'JACKET',
       'PANTS', 'HAT', 'GLOVES', 'TIE', 'SKIRT', 'STOCKINGS', 'SHIRT',
       'UNDERPANTS', 'JEANS', 'SUIT', 'SHOES'], dtype=object)

##### As seen above, COLOR and CATEGORY are within the Length requirements. 

#### 2) Check Unique Constraints, missing values and determine primary /composite keys

In [300]:
# count distinct values in ITEM field 
items_prep.ITEM.nunique()

10944

#### Reporting: As seen above, ITEM field is not a primary key. In fact, ITEM field could be a foreign key to TransactionDB; COLOR and CATEGORY is distinct to each ITEM code. 

In [301]:
# number of COLOR
items_prep.COLOR.value_counts().count()

5

In [302]:
# number of categories 
items_prep.CATEGORY.value_counts().count()

18

#### Determining primary key or composite key field(s)

In [309]:
# find number of distinct SUPLID
items_prep.SUPLID.nunique()

25

#### check missing values

In [308]:
items_prep.notnull().sum()

ITEM             284390
CATEGORY         284390
COLOR            284390
SUPLID           273450
PURCHASEPRICE    284390
SALEPRICE        284390
dtype: int64

In [305]:
print('' in items_prep['SUPLID'])

False


In [306]:
print('null' in items_prep['SUPLID'])

False


In [307]:
# Check counts of missing values
items_prep.isnull().sum()*100/len(items_prep)

ITEM             0.00000
CATEGORY         0.00000
COLOR            0.00000
SUPLID           3.84683
PURCHASEPRICE    0.00000
SALEPRICE        0.00000
dtype: float64

#### Reporting: As seen above, there are 3.84% of missing data in SUPLID. 

In [310]:
# count the number of missing values in PURCHASEPRICE
items_prep[items_prep.PURCHASEPRICE == 0.0].count()

ITEM             26
CATEGORY         26
COLOR            26
SUPLID           25
PURCHASEPRICE    26
SALEPRICE        26
dtype: int64

In [311]:
# count the % of missing values in PURCHASEPRICE
items_prep[items_prep.PURCHASEPRICE == 0.0].count()*100/len(items_prep)

ITEM             0.009142
CATEGORY         0.009142
COLOR            0.009142
SUPLID           0.008791
PURCHASEPRICE    0.009142
SALEPRICE        0.009142
dtype: float64

#### As seen above, there are 0.009 % of missing PURCHASEPRICE (n=200) 

In [312]:
items_prep[items_prep.PURCHASEPRICE == 0.0].groupby(['SALEPRICE']).count()

ITEM  CATEGORY  COLOR  SUPLID  PURCHASEPRICE
SALEPRICE                                              
0.0          26        26     26      25             26

#### As seen above, there are 0.009 % of missing SALEPRICE (n=200) 

#### Reporting: 1)  ITEM is a foreign Key. Each item refers to distinct COLOR and CATEGORY. The same ITEM can have multiple SUPLID.  2) No missing values except those in SUPLID field. 

### Decompose ITEMDB

In [318]:
supls = pd.DataFrame(items_prep, columns = ['SUPLID', 'PURCHASEPRICE', 'ITEM' ])
supls.head()

SUPLID  PURCHASEPRICE    ITEM
0  41560EE        1867.09  127521
1  16091ZZ        1862.08  151662
2  11370LL         914.81  453756
3  41560EE        1913.00  434386
4  48094JJ         851.43  422712

In [340]:
supls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 284390 entries, 0 to 2180594
Data columns (total 3 columns):
SUPLID           273450 non-null object
PURCHASEPRICE    284390 non-null float64
ITEM             284390 non-null int64
dtypes: float64(1), int64(1), object(1)
memory usage: 18.7+ MB


In [327]:
supls.SUPLID.unique()

array(['41560EE', '16091ZZ', '11370LL', '48094JJ', '41560MM', '11370TT',
       '42596VV', '43136JJ', '42258KK', nan, '46396EE', '41560GG',
       '18767CC', '46848VV', '42596PP', '16091DD', '46848AA', '48094CC',
       '42258WW', '13205GG', '18767GG', '14656DD', '46396XX', '43136TT',
       '14656UU', '13205VV'], dtype=object)

In [339]:
supls.isnull().sum()

SUPLID           10940
PURCHASEPRICE        0
ITEM                 0
dtype: int64

In [333]:
supls.SUPLID.value_counts()

48094JJ    10944
41560MM    10941
11370LL    10940
46848VV    10940
42258KK    10940
41560GG    10940
48094CC    10940
18767CC    10940
42258WW    10939
43136TT    10939
13205GG    10939
46396EE    10939
18767GG    10938
13205VV    10937
14656DD    10937
11370TT    10937
16091ZZ    10937
46848AA    10937
14656UU    10937
16091DD    10937
43136JJ    10936
42596VV    10935
41560EE    10935
46396XX    10933
42596PP    10933
Name: SUPLID, dtype: int64

In [319]:
supls.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 284390 entries, 0 to 2180594
Data columns (total 3 columns):
SUPLID           273450 non-null object
PURCHASEPRICE    284390 non-null float64
ITEM             284390 non-null int64
dtypes: float64(1), int64(1), object(1)
memory usage: 18.7+ MB


In [322]:
supls.SUPLID[supls.SUPLID==' '].count()

0

In [350]:
#### Save a copy of items pre-processed data 
items_prep = pd.DataFrame(items)
items_prep.to_csv('items_prep.csv')
items_prep.to_csv(r'C:\Users\YUJI6004\items_prep.csv', index=False)
# load the saved copy and verify dtypes consistency 
items_prep = pd.read_csv('items_prep.csv', sep =',')
items_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 6 columns):
ITEM             int64
CATEGORY         object
COLOR            object
SUPLID           object
PURCHASEPRICE    float64
SALEPRICE        float64
dtypes: float64(2), int64(1), object(3)
memory usage: 100.2+ MB


### Customers dataset

#### 1) Check range constraints of USERID and GENDER fields

In [313]:
#check statistics of quantitative variable 
customers.describe()

USERID
count    574.000000
mean   29972.229965
std    11550.527178
min    10060.000000
25%    20047.500000
50%    29380.000000
75%    39952.500000
max    49980.000000

In [71]:
# check data integrity of GENDER.
customers.GENDER.unique()

array(['F', 'M', 'N'], dtype=object)

#### Reporting: As seen above, USERID is within the required range [10000, 50000]. GENDER field meets with data constraints: [F, M,N]

#### 2) Check Primary/Foreign keys

In [69]:
#count the distinct number of USERID
customers.USERID.nunique()

574


#### Reporting: As seen above, the count of total rows is the same as the number of unique values in USERIDs (=574). Thus, USERID is the primary key 

#### 3) Check missing values

In [194]:
# check missing values in customers DB
customers.isna().any()

USERID       False
GENDER       False
DOB          False
COUNTRY      False
EDUCATION    False
HOBBY        False
dtype: bool

In [72]:
# count the number each GENDER 
customers.GENDER.value_counts()

F    304
M    214
N     56
Name: GENDER, dtype: int64

#### As seen above, no missing values in the customer DB. However, it is worth mentiong that the GENDER field has 56 N USERIDs. Those are the USERs without gender info, perhaps their gender identity is not disclosed or considered as transgender, or representing other minor sexualities  etc. 

In [73]:
# check unique values in COUNTRY
customers.COUNTRY.unique()

array(['Ethiopia', 'United Arab Emirates', 'Martinique', 'Germany',
       'Madagascar', 'Guinea', 'Japan', 'Algeria', 'Puerto Rico',
       'Antarctica', 'Senegal', 'France', 'Chile', 'Gibraltar', 'Reunion',
       'Morocco', 'Mexico', 'Uruguay', 'Guernsey', 'Macedonia', 'Liberia',
       'Jordan', 'Malta', 'Bahamas', 'Greenland', 'Cyprus', 'Mozambique',
       'Sweden', 'Thailand', 'Uganda', 'South Africa', 'Andorra',
       'Croatia', 'Barbados', 'Moldova', 'Faroe Islands', 'Romania',
       'Italy', 'China', 'Venezuela', 'Hungary', 'United Kingdom',
       'Rwanda', 'Belgium', 'South Korea'], dtype=object)

#### Reporting: as shown above, no missing values in COUNTRY

In [74]:
# check unique values in EDUCATION
customers.EDUCATION.unique()

array([' ', 'Bachelor', 'Master', 'High school', 'Less than high school'],
      dtype=object)

In [193]:
customers.EDUCATION.value_counts(ascending=True)

Master                   106
Null                     110
Bachelor                 112
Less than high school    117
High school              129
Name: EDUCATION, dtype: int64

In [190]:
customers.EDUCATION.value_counts(normalize=True)

High school              0.224739
Less than high school    0.203833
Bachelor                 0.195122
Null                     0.191638
Master                   0.184669
Name: EDUCATION, dtype: float64

##### As seen above, 19.16% of Education data is missing 

#### It is worth mentioning that ' ' is stored as a string, thus the isnull() function is evaluated as false.

In [75]:
# replace ' ' with Null in EDUCATION
customers.EDUCATION = customers.EDUCATION.replace(to_replace = ' ', value = 'Null')

In [183]:
# Verify the replacement 
customers.EDUCATION.unique()

array(['Null', 'Bachelor', 'Master', 'High school',
       'Less than high school'], dtype=object)

In [207]:
# check unique values in HOBBY
customers.HOBBY.unique()

array(['Laser tag', 'Ultra running', 'Pottery', 'Hamsters', 'Other',
       'Running', 'Cheerleading', 'Beekeeping', 'Macrame', 'Kayaking',
       ' ', 'Meteorology', 'Paintball', 'Embroidery', 'Painting',
       'Puzzles', 'Hunting', 'Horseback riding', 'Flying', 'Climbing',
       'Geocaching', 'Saxophone', 'Sculpting', 'Origami', 'Gaming'],
      dtype=object)

In [208]:
# check number of missing values in HOBBY
customers.HOBBY.value_counts(ascending=True)

Beekeeping          15
Paintball           16
Geocaching          17
Cheerleading        17
                    18
Macrame             20
Kayaking            20
Puzzles             20
Horseback riding    20
Gaming              21
Saxophone           23
Climbing            23
Meteorology         23
Running             23
Pottery             24
Sculpting           24
Hunting             25
Painting            25
Flying              25
Ultra running       28
Hamsters            28
Other               28
Laser tag           29
Origami             30
Embroidery          32
Name: HOBBY, dtype: int64

In [211]:
# replace ' ' with Null in HOBBY
customers.HOBBY = customers.HOBBY.replace(to_replace = ' ', value = 'Null')

In [212]:
# Verify the replacement 
customers.EDUCATION.unique()

array(['Null', 'Bachelor', 'Master', 'High school',
       'Less than high school'], dtype=object)

In [210]:
# check % of missing values in HOBBY
customers.HOBBY.value_counts(normalize=True)

Embroidery          0.055749
Origami             0.052265
Laser tag           0.050523
Other               0.048780
Hamsters            0.048780
Ultra running       0.048780
Flying              0.043554
Painting            0.043554
Hunting             0.043554
Sculpting           0.041812
Pottery             0.041812
Running             0.040070
Meteorology         0.040070
Climbing            0.040070
Saxophone           0.040070
Gaming              0.036585
Horseback riding    0.034843
Puzzles             0.034843
Kayaking            0.034843
Macrame             0.034843
                    0.031359
Cheerleading        0.029617
Geocaching          0.029617
Paintball           0.027875
Beekeeping          0.026132
Name: HOBBY, dtype: float64

In [202]:
customers.sort_values('DOB',ascending = True)

USERID GENDER        DOB               COUNTRY              EDUCATION  \
55    36420      N 1965-01-03               Moldova                   Null   
541   42170      F 1965-01-10             Greenland                   Null   
159   27840      F 1965-01-20             Gibraltar                 Master   
425   37500      M 1965-02-10           South Korea                 Master   
242   17190      M 1965-02-16                Guinea                 Master   
290   42430      F 1965-04-10               Uruguay                   Null   
402   49260      F 1965-05-13               Liberia            High school   
246   15970      M 1965-05-24  United Arab Emirates            High school   
366   48220      N 1965-08-07                Guinea                   Null   
554   23890      M 1965-08-11               Belgium            High school   
547   49000      N 1965-09-26               Senegal  Less than high school   
371   12080      M 1965-10-15               Hungary  Less than high school   
249   16160      M 1965-10-29                 Italy            High school   
64    13560      F 1966-01-26                 Malta                 Master   
34    21850      F 1966-02-20                 Chile               Bachelor   
9     37480      M 1966-04-09            Antarctica                   Null   
527   21970      F 1966-04-16               Moldova  Less than high school   
52    35110      M 1966-05-05               Reunion               Bachelor   
407   39360      M 1966-07-29              Thailand                 Master   
350   49100      F 1966-08-06                Cyprus               Bachelor   
165   21960      F 1966-09-12                 Italy  Less than high school   
533   46000      M 1966-10-06         Faroe Islands  Less than high school   
276   21830      M 1966-11-16              Barbados                   Null   
141   26100      F 1967-02-28            Martinique                 Master   
28    46530      F 1967-03-24                 Japan                   Null   
418   47210      M 1967-04-08              Guernsey                   Null   
409   25850      F 1967-04-27             Macedonia                   Null   
521   42570      F 1967-05-08               Algeria            High school   
562   23910      M 1967-05-10            Mozambique                 Master   
560   30720      F 1967-05-11                Cyprus                 Master   
..      ...    ...        ...                   ...                    ...   
540   33940      M 2000-03-30            Martinique            High school   
482   22830      F 2000-04-20               Romania  Less than high school   
72    18430      F 2000-04-27                Mexico                   Null   
113   29020      F 2000-05-15                Mexico               Bachelor   
111   28600      F 2000-07-15             Gibraltar  Less than high school   
261   36610      F 2000-07-16               Algeria                   Null   
317   46700      M 2000-07-17               Andorra  Less than high school   
435   45580      F 2000-07-20              Ethiopia            High school   
537   13170      F 2000-09-10             Greenland  Less than high school   
525   41590      M 2000-09-14               Croatia               Bachelor   
222   32950      F 2000-09-18                Jordan               Bachelor   
325   36950      F 2000-11-17             Greenland                 Master   
211   35430      F 2000-11-21  United Arab Emirates               Bachelor   
305   25460      M 2000-12-13          South Africa               Bachelor   
8     27260      F 2000-12-28           Puerto Rico            High school   
570   24340      F 2001-02-04               Moldova                 Master   
486   40810      M 2001-02-16               Morocco  Less than high school   
47    15770      F 2001-03-05                Guinea            High school   
108   14730      F 2001-05-26               Reunion  Less than high school   
83    33220      F 2001-06-

##### Transactions dataset

In [270]:
transactions.head()

USERID WEBBROWSER PPC_ADD    ITEM PURCHASE    QTY  DISCOUNT    PAYMENT  \
0   15410     CHROME   KEVIN  309261      YES  15410        28  APPLE PAY   
1   33510   EXPLORER   SENDI  484592      YES  33510        30              
2   26390  MOBILEAPP   KEVIN  191104      YES  26390         4              
3   48580  MOBILEAPP   SENDI  124180      YES  48580        29       AMEX   
4   14600  MOBILEAPP   KEVIN  223591      YES  14600         7  APPLE PAY   

  WAREHOUSE  SHIPDAYS DELIVERYDATE REVIEW  RATING  TRACKNO TIMESTAMP  
0         D         6   2018-05-25     NO       1   348721  21:54:45  
1         D         5   2018-05-01     NO       4   373987  05:44:25  
2         B         7   2018-07-07    YES       0   421013  06:58:15  
3         E         5   2018-02-13     NO       3   406643  12:50:23  
4         B         6   2018-01-10     NO       0   365623  05:23:37

#### To start with, let's look at the quantitative variables 

In [79]:
#check statistics of quantitative variable 
transactions.describe()

USERID          ITEM           QTY      DISCOUNT      SHIPDAYS  \
count  2.188800e+06  2.188800e+06  2.188800e+06  2.188800e+06  2.188800e+06   
mean   3.000972e+04  2.983991e+05  3.000972e+04  1.457914e+01  4.596594e+00   
std    1.156518e+04  1.152450e+05  1.156518e+04  8.974733e+00  3.039147e+00   
min    1.006000e+04  1.000130e+05  1.006000e+04  0.000000e+00  0.000000e+00   
25%    2.003000e+04  1.995800e+05  2.003000e+04  7.000000e+00  2.000000e+00   
50%    2.945000e+04  2.968850e+05  2.945000e+04  1.400000e+01  4.000000e+00   
75%    3.997000e+04  3.975310e+05  3.997000e+04  2.200000e+01  7.000000e+00   
max    5.000000e+04  4.999970e+05  5.000000e+04  3.000000e+01  1.000000e+01   

             RATING       TRACKNO               TIMESTAMP  
count  2.188800e+06  2.188800e+06                 2188800  
mean   2.223496e+00  3.999524e+05  0 days 09:54:37.335662  
std    1.791621e+00  5.772411e+04  0 days 06:42:45.117814  
min    0.000000e+00  3.000000e+05         0 days 00:00:00  
25%    0.000000e+00  3.499650e+05         0 days 05:23:37  
50%    2.000000e+00  4.000260e+05         0 days 10:00:05  
75%    4.000000e+00  4.499980e+05         0 days 12:50:23  
max    5.000000e+00  5.000000e+05         0 days 23:41:05

#### As seen above, the summary statistics of USERID and QTY is identical!! Thus, QTY is not a reliable field. 

In [272]:
# count the number of distinct values in USERID
transactions.USERID.nunique()

575

In [81]:
# count the number of distinct values in ITEM
transactions.ITEM.nunique()

10944

In [82]:
# count the number of distinct values in QTY
transactions.QTY.nunique()

575

In [269]:
transactions.sort_values('QTY', axis=0, ascending = True, na_position ='first')

USERID WEBBROWSER   PPC_ADD    ITEM PURCHASE    QTY  DISCOUNT  \
2060360   10060  MOBILEAPP    BIANCA  227335      YES  10060        30   
754547    10060   EXPLORER     KEVIN  371349      YES  10060        19   
220058    10060  MOBILEAPP     KEVIN  485190      YES  10060        30   
1518391   10060  MOBILEAPP        NO  116133      YES  10060         6   
195986    10060  MOBILEAPP     SENDI  330373           10060        18   
1786679   10060     CHROME        NO  388899      YES  10060        30   
1057156   10060     CHROME     KEVIN  198228      YES  10060        23   
1481543   10060  MOBILEAPP        NO  243442      YES  10060         6   
1593234   10060     CHROME  ISABELLA  106474      YES  10060        23   
1185166   10060     CHROME     KENZA  338732      YES  10060        19   
1450184   10060  MOBILEAPP  ISABELLA  154431      YES  10060        18   
1481616   10060     CHROME     KEVIN  199203      YES  10060        30   
137304    10060     SAFARI  ISABELLA  283456      YES  10060        30   
1853440   10060  MOBILEAPP        NO  269845      YES  10060        30   
2071527   10060     CHROME        NO  125922       NO  10060        24   
1258012   10060     CHROME     KEVIN  182433      YES  10060        28   
1185018   10060     SAFARI     KENZA  116471      YES  10060        19   
1969610   10060                   NO  193288      YES  10060        28   
1899515   10060     SAFARI     KEVIN  119812      YES  10060        10   
998473    10060     CHROME    BIANCA  392695      YES  10060        26   
410976    10060     CHROME     SENDI  479899       NO  10060         4   
3379      10060   EXPLORER  ISABELLA  348768      YES  10060        28   
1517922   10060     CHROME  ISABELLA  325004           10060        20   
1786043   10060  MOBILEAPP     KEVIN  313083      YES  10060         2   
1660846   10060     CHROME        NO  357491      YES  10060        26   
196235    10060     CHROME     KEVIN  211826      YES  10060        19   
245695    10060     CHROME    BIANCA  304503      YES  10060        28   
1969629   10060  MOBILEAPP     KEVIN  250046       NO  10060         3   
1100406   10060     CHROME        NO  141379      YES  10060        18   
1744645   10060     CHROME        NO  360559      YES  10060         6   
...         ...        ...       ...     ...      ...    ...       ...   
1823357   50000     SAFARI     KEVIN  464312      YES  50000        23   
1703555   50000     CHROME            236201       NO  50000         4   
1456890   50000                KEVIN  201829      YES  50000        28   
1934808   50000  MOBILEAPP    BIANCA  375483      YES  50000        30   
1976299   50000     CHROME     KENZA  404616      YES  50000         1   
680652    50000     CHROME  ISABELLA  164519      YES  50000         5   
2135844   50000  MOBILEAPP  ISABELLA  218248       NO  50000        23   
1408128   50000     CHROME     KEVIN  365564       NO  50000        29   
1479219   50000  MOBILEAPP     KEVIN  315137      YES  50000        29   
2169268   50000     SAFARI    BIANCA  481277      YES  50000         1   
2024191   50000  MOBILEAPP     KEVIN  296872           50000         1   
710755    50000                   NO  267219      YES  50000         4   
299991    50000   EXPLORER  ISABELLA  213412      YES  50000         1   
2159159   50000  MOBILEAPP    BIANCA  412793      YES  50000        23   
1985640   50000     CHROME    BIANCA  236500      YES  50000         6   
1230698   50000     SAFARI     KEVIN  464806      YES  50000        16   
56399     50000   EXPLORER     KENZA  142861      YES  50000        20   
1746883   50000   EXPLORER  ISABELLA  229168      YES  50000         6   
188056    50000               BIANCA  448712           50000        16   
20009     50000     SAFARI  ISABELLA  359116      YES  50000         7   
605790    50000     SAFARI  ISABELLA  180470      YES  50000         8   
817361    50000  MOBILEAPP     SENDI  235850       NO  50000        23   
164565    500

In [268]:
transactions[transactions.QTY == 0].count()

USERID          0
WEBBROWSER      0
PPC_ADD         0
ITEM            0
PURCHASE        0
QTY             0
DISCOUNT        0
PAYMENT         0
WAREHOUSE       0
SHIPDAYS        0
DELIVERYDATE    0
REVIEW          0
RATING          0
TRACKNO         0
TIMESTAMP       0
dtype: int64

In [83]:
# Number count of distinct values in RATING 
transactions.RATING.value_counts()

4    567549
0    566753
2    405571
1    324295
5    243629
3     81003
Name: RATING, dtype: int64

In [250]:
transactions.RATING.value_counts(normalize = True)

4    0.259297
0    0.258933
2    0.185294
1    0.148161
5    0.111307
3    0.037008
Name: RATING, dtype: float64

#### As seen above, 25.98% of data missing in RATING

In [84]:
# Checking null values in dataframe 
transactions.isnull().any()

USERID          False
WEBBROWSER      False
PPC_ADD         False
ITEM            False
PURCHASE        False
QTY             False
DISCOUNT        False
PAYMENT         False
WAREHOUSE       False
SHIPDAYS        False
DELIVERYDATE    False
REVIEW          False
RATING          False
TRACKNO         False
TIMESTAMP       False
dtype: bool

In [85]:
#value count by % in TIMESTAMP
transactions['TIMESTAMP'].value_counts(normalize=True)

02:08:26    0.016907
12:25:23    0.016868
07:56:56    0.016816
05:56:53    0.016811
12:03:51    0.016809
05:44:25    0.016799
12:18:21    0.016786
10:30:09    0.016775
00:28:54    0.016771
09:42:18    0.016770
05:23:37    0.016755
00:00:00    0.016747
21:54:45    0.016737
11:56:58    0.016734
12:07:37    0.016728
21:09:24    0.016725
20:11:54    0.016723
23:16:25    0.016721
12:09:28    0.016717
12:50:23    0.016708
20:18:53    0.016705
06:16:02    0.016700
09:42:39    0.016699
13:24:06    0.016699
09:54:58    0.016696
05:50:52    0.016683
09:02:29    0.016668
12:48:46    0.016668
11:23:13    0.016666
01:47:18    0.016664
12:16:02    0.016662
11:50:25    0.016653
12:25:54    0.016649
00:03:15    0.016642
07:02:59    0.016641
07:05:34    0.016640
19:02:37    0.016625
02:47:02    0.016624
11:09:59    0.016624
01:13:51    0.016622
13:57:21    0.016614
00:44:26    0.016611
06:58:15    0.016606
09:09:58    0.016605
20:16:01    0.016603
01:45:36    0.016598
01:26:31    0.016596
02:31:39    0

##### As shown above, TIMESTAMP hour is within [00, 24].

#### Let's look at the categorical values in transactions DB

In [87]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2188800 entries, 0 to 2188799
Data columns (total 15 columns):
USERID          int64
WEBBROWSER      object
PPC_ADD         object
ITEM            int64
PURCHASE        object
QTY             int64
DISCOUNT        int64
PAYMENT         object
WAREHOUSE       object
SHIPDAYS        int64
DELIVERYDATE    datetime64[ns]
REVIEW          object
RATING          int64
TRACKNO         int64
TIMESTAMP       timedelta64[ns]
dtypes: datetime64[ns](1), int64(7), object(6), timedelta64[ns](1)
memory usage: 250.5+ MB


In [273]:
transactions.PAYMENT.value_counts()

                   438436
VISA/MASTERCARD    438072
PAYPAL             437730
APPLE PAY          437380
AMEX               437182
Name: PAYMENT, dtype: int64

In [274]:
transactions.PAYMENT.value_counts()*100/len(transactions)

                   20.030885
VISA/MASTERCARD    20.014254
PAYPAL             19.998629
APPLE PAY          19.982639
AMEX               19.973593
Name: PAYMENT, dtype: float64

#### As shown, 20% of missing data in PAYMENT

In [276]:
transactions.WAREHOUSE.unique()

array(['D', 'B', 'E', ' ', 'C', 'A'], dtype=object)

In [277]:
transactions.WAREHOUSE.value_counts()

A    365929
C    365246
D    364959
     364427
E    364367
B    363872
Name: WAREHOUSE, dtype: int64

In [278]:
transactions.WAREHOUSE.value_counts()*100/len(transactions)

A    16.718247
C    16.687043
D    16.673931
     16.649625
E    16.646884
B    16.624269
Name: WAREHOUSE, dtype: float64

##### As seen above, 16.65% of WAREHOUSE data is missing (n=364427)

In [86]:
#value count TIMESTAMP (order=ascending)
transactions['TIMESTAMP'].value_counts(ascending=True)

01:31:40    36021
02:30:52    36174
13:09:53    36179
10:00:05    36182
23:41:05    36208
15:18:22    36216
01:18:40    36232
05:55:03    36234
00:31:21    36237
15:20:04    36245
23:07:53    36295
19:34:14    36324
02:31:39    36324
01:26:31    36325
01:45:36    36330
20:16:01    36341
09:09:58    36346
06:58:15    36348
00:44:26    36359
13:57:21    36364
01:13:51    36383
11:09:59    36386
02:47:02    36386
19:02:37    36389
07:05:34    36422
07:02:59    36424
00:03:15    36425
12:25:54    36442
11:50:25    36450
12:16:02    36470
01:47:18    36475
11:23:13    36478
12:48:46    36482
09:02:29    36483
05:50:52    36516
09:54:58    36545
13:24:06    36550
09:42:39    36551
06:16:02    36554
20:18:53    36564
12:50:23    36571
12:09:28    36591
23:16:25    36598
20:11:54    36603
21:09:24    36607
12:07:37    36615
11:56:58    36627
21:54:45    36634
00:00:00    36655
05:23:37    36673
09:42:18    36707
00:28:54    36708
10:30:09    36718
12:18:21    36742
05:44:25    36770
12:03:51  

In [88]:
transactions.WEBBROWSER.unique()

array(['CHROME', 'EXPLORER', 'MOBILEAPP', 'SAFARI', ' '], dtype=object)

In [89]:
transactions['WEBBROWSER'].value_counts(normalize = True)

MOBILEAPP    0.399772
CHROME       0.333050
SAFARI       0.133532
EXPLORER     0.066856
             0.066791
Name: WEBBROWSER, dtype: float64

In [263]:
transactions['WEBBROWSER'].value_counts(ascending = True)

             146192
EXPLORER     146334
SAFARI       292274
CHROME       728979
MOBILEAPP    875021
Name: WEBBROWSER, dtype: int64

#### As displayed, 0.0668 % of missing data (n=146192) in WEBBROWSER.

In [259]:
transactions['REVIEW'].value_counts(ascending = True)

        219272
YES     876310
NO     1093218
Name: REVIEW, dtype: int64

In [255]:
transactions['REVIEW'].value_counts(normalize = True)

NO     0.499460
YES    0.400361
       0.100179
Name: REVIEW, dtype: float64

##### As shown above, 10% of missing data in REVIEW (n=219,272)

In [261]:
transactions.PPC_ADD.value_counts()

KEVIN       776220
BIANCA      465492
ISABELLA    310562
NO          310555
KENZA       155264
SENDI       155225
             15482
Name: PPC_ADD, dtype: int64

In [262]:
transactions.PPC_ADD.value_counts()*100/len(transactions)

KEVIN       35.463268
BIANCA      21.266996
ISABELLA    14.188688
NO          14.188368
KENZA        7.093567
SENDI        7.091785
             0.707328
Name: PPC_ADD, dtype: float64

#### As shown, 0.7% of missing data in PPC_ADD

In [264]:
transactions.PURCHASE.value_counts()

YES    1531405
NO      438482
        218913
Name: PURCHASE, dtype: int64

In [265]:
transactions.PURCHASE.value_counts()*100/len(transactions)

YES    69.965506
NO     20.032986
       10.001508
Name: PURCHASE, dtype: float64

#### As seen above, 10% of missing data in PURCHASE (n=218913)

In [285]:
list(transactions.columns.values)

['USERID',
 'WEBBROWSER',
 'PPC_ADD',
 'ITEM',
 'PURCHASE',
 'QTY',
 'DISCOUNT',
 'PAYMENT',
 'WAREHOUSE',
 'SHIPDAYS',
 'DELIVERYDATE',
 'REVIEW',
 'RATING',
 'TRACKNO',
 'TIMESTAMP']

In [288]:
# check duplicates
transactions.duplicated(['USERID',
 'WEBBROWSER',
 'PPC_ADD',
 'ITEM',
 'PURCHASE',
 'DISCOUNT',
 'PAYMENT',
 'WAREHOUSE',
 'SHIPDAYS',
 'DELIVERYDATE',
 'REVIEW',
 'RATING',
 'TRACKNO',
 'TIMESTAMP']).value_counts()* 100/len(transactions)

False    100.0
dtype: float64

### SECTION III: DATA EXPLORATION 

#### Part III.1.A - ITEMS

In [91]:
items_prep.corr()

ITEM  PURCHASEPRICE  SALEPRICE
ITEM           1.000000      -0.003493  -0.004547
PURCHASEPRICE -0.003493       1.000000   0.917953
SALEPRICE     -0.004547       0.917953   1.000000